src/Polysemy/Fresh.hs

```diff
@@ -0,0 +1,115 @@
+--
+-- __Beware:__ This is safe only if:
+--
+--   1. This is run after all interpreters which may revert local state
+--      or produce multiple, inconsistent instances of local state.
+--      This includes interpreters that may backtrack or produce multiple results
+--      (such as 'Polysemy.Error.runError' or 'Polysemy.NonDet.runNonDet').
+--
+--   2. You don't use any interpreter which may cause the final monad
+--      to revert local state or produce multiple, inconsistent instances of local
+--      state. This includes certain 'Polysemy.Final.Final'/@lower-@ interpeters
+--      such as 'Polysemy.Error.lowerError' or 'Polysemy.Final.MTL.errorToFinal',
+--      as well as interpreters for 'Polysemy.Async.Async'.
+--
+-- Prefer 'freshToIO' whenever possible.
+-- If you can't use 'runFreshUnsafe' safely, nor use 'freshToIO', consider
+-- 'runFreshUnsafePerformIO'.
+runFreshUnsafe :: Sem (Fresh Integer ': r) a
+               -> Sem r a
+runFreshUnsafe =
+    (fmap snd .)
+  $ (runState @Integer 0 .)
+  $ reinterpret
+  $ \Fresh -> do
+    s <- get
+    put $! s + 1
+    return s
+{-# INLINE runFreshUnsafe #-}
+
+-----------------------------------------------------------------------------
+-- | Runs a 'Fresh' effect through generating 'Unique's using
+-- 'unsafePerformIO'.
+--
+-- Ironically, despite the fact that this uses 'unsafePerformIO', and
+-- 'runFreshUnsafe' uses no unsafe operations whatsoever, this is still
+-- typically safer to use than 'runFreshUnsafe', although 'runFreshUnsafe'
+-- is perhaps more efficient.
+--
+-- The worst thing that this particular use of 'unsafePerformIO' could result
+-- in is the loss of referential transparency, as rerunning an interpreter using
+-- 'runFreshUnsafePerformIO' will create different 'Unique's. This should
+-- never matter.
+--
+-- This could be potentially be less efficient than 'runFreshUnsafe'.
+--
+-- If you ever observe that multiple invocations of 'fresh' produce the same
+-- 'Unique' under 'runFreshUnsafePerformIO', then open an issue over at the
+-- GitHub repository.
+runFreshUnsafePerformIO :: Sem (Fresh Unique ': r) a
+                        -> Sem r a
+runFreshUnsafePerformIO = usingSem $ \u ->
+  case decomp u of
+    Right (Weaving Fresh s _ ex _) -> do
+      let !uniq = unsafePerformIO (newUnique' u)
+          {-# NOINLINE uniq #-}
+      pure $ ex (uniq <$ s)
+    Left g -> liftSem (hoist runFreshUnsafePerformIO g)
+-- KingoftheHomeless: I've tried very hard to prevent optimizations from
+-- sharing the call to 'unsafePerformIO'.
+-- The inlining of 'interpret' is so that I can give
+-- 'u' to 'newUnique'', and thus prevent 'uniq' from floating outside the
+-- lambda. This interpreter might even be safe to INLINE, but I'm not taking
+-- any chances.
+{-# NOINLINE runFreshUnsafePerformIO #-}
+
+newUnique' :: Union (Fresh Unique ': r) (Sem (Fresh Unique ': r)) a -> IO Unique
+newUnique' (Union _ _) = newUnique
+{-# NOINLINE newUnique' #-}
```

Comment:
I'd love to see a `runFreshNum :: Num n => Sem (Fresh n ': r) a -> Sem r a` interpretation here

Reply:
Are you in favour of generalizing `runFreshUnsafe` to have that signature, or should we have a seperate `runFreshNumUnsafe`? The issue with generalizing `runFreshUnsafe` is that you'd run into ambiguity issues if you don't care about the specific `Num`.

Reply:
I don't have a strong preference. But I use this effect a lot for generating user-facing symbols, which are definitely nicer if they're monotonic. I don't think the ambiguity will be an issue for any user-facing code that uses `Fresh`.

Reply:
Perhaps what we could do is have `runFreshUnsafe` use a newtype of `Integer` instead (just like `Unique`), so that can be used to provide a generic user-facing symbol, and have `runFreshNumUnsafe` seperate.

Reply:
Yeah, I like that idea. I'll implement that.

Reply:
How do you feel about `runFreshEnumUnsafe` instead of `runFreshNumUnsafe`?

Reply:
Love it. Make it so!